### PR TITLE
[Fixes #7529] GeoNode V2 API layers, maps and resources responses do not include URLs for each resource

### LINKED ISSUE #7529: GeoNode V2 API layers, maps and resources responses do not include URLs for each resource
## Expected Behavior
`link` property should be included in the response for resources, containing the link to the API detail endpoint.

## Actual Behavior
Given
1. `/api/v2/layers/` - returns a list of layers
2. `/api/v2/layers/{layer-id}/` - returns the details of a single layer

The current implementation of the V2 API does not provide a URL for going from 1. to 2. - the client needs to guess the URL.

## Steps to Reproduce the Problem

  1. https://master.demo.geonode.org/api/v2/resources

## Specifications

  - GeoNode version: master
  - Installation method (manual, GeoNode Docker, SPCGeoNode Docker):
  - Platform:
  - Additional details:

### PR DESCRIPTION
**Pull request by marthamareal**
<Include a few sentences describing the overall goals for this Pull Request>
References  #7529

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

File: geonode/base/api/serializers.py [hunk 1]
```diff
@@ -49,6 +53,18 @@
 logger = logging.getLogger(__name__)
 
 
+class BaseDynamicModelSerializer(DynamicModelSerializer):
+
+    def to_representation(self, instance):
+        data = super(BaseDynamicModelSerializer, self).to_representation(instance)
+        path = reverse(self.Meta.view_name)
+        if not path.endswith('/'):
+            path = path + '/'
```
Comment (afabiani) [inline geonode/base/api/serializers.py:62]: Please use f-stings instead of concatenation

File: geonode/base/api/serializers.py [hunk 2]
```diff
@@ -49,6 +53,18 @@
 logger = logging.getLogger(__name__)
 
 
+class BaseDynamicModelSerializer(DynamicModelSerializer):
+
+    def to_representation(self, instance):
+        data = super(BaseDynamicModelSerializer, self).to_representation(instance)
+        path = reverse(self.Meta.view_name)
```
Comment (afabiani) [inline geonode/base/api/serializers.py:60]: Please put a check here in the case the "view_name" does not exist

**Review by afabiani — CHANGES_REQUESTED** [review]
Add test cases please

**Comment by marthamareal** [comment]
@afabiani I have updated the PR

File: geonode/base/api/serializers.py [hunk 3]
```diff
@@ -49,6 +53,21 @@
...
 
+class BaseDynamicModelSerializer(DynamicModelSerializer):
+
+    def to_representation(self, instance):
+        data = super(BaseDynamicModelSerializer, self).to_representation(instance)
+        try:
+            path = reverse(self.Meta.view_name)
+            if not path.endswith('/'):
+                path = f"{path}/"
+                url = urljoin(path, str(instance.pk))
```
Comment (afabiani) [inline geonode/base/api/serializers.py:64]: I guess this is overintented

**Review by afabiani — CHANGES_REQUESTED** [review]

File: geonode/base/api/serializers.py (hunk 3 above)
Comment (giohappy) [reply geonode/base/api/serializers.py:64]: @afabiani the problem is that urljoin will consider the first slash it finds in the first path. So it path misses the final slash the resulting url might be truncated.

Comment (afabiani) [reply geonode/base/api/serializers.py:64]: @giohappy there's a syntax error. Like it is now the url and data[link] will be populated only if `not path.endswith('/')`

Comment (marthamareal) [reply geonode/base/api/serializers.py]: updated

Comment (giohappy) [reply geonode/base/api/serializers.py]: oh right @afabiani I thought you were referring to the extra checl.

**Review by afabiani — APPROVED** [review]

**Comment by afabiani** [comment]
@marthamareal there are api tests failing

```
FAIL: test_geoapps_list (geonode.geoapps.api.tests.BaseApiTests)
Ensure we can access the GeoApps list.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/geonode/geonode/geoapps/api/tests.py", line 152, in test_geoapps_list
    self.assertEqual(
AssertionError: '{"test_data": {"test": ["test_1","test_2","test_3"]}}' != {'test_data': {'test': ['test_1', 'test_2', 'test_3']}}

----------------------------------------------------------------------
Ran 23 tests in 163.573s
```

**Comment by marthamareal** [comment]
> @marthamareal there are api tests failing
> 
> ```
> FAIL: test_geoapps_list (geonode.geoapps.api.tests.BaseApiTests)
> Ensure we can access the GeoApps list.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/usr/src/geonode/geonode/geoapps/api/tests.py", line 152, in test_geoapps_list
>     self.assertEqual(
> AssertionError: '{"test_data": {"test": ["test_1","test_2","test_3"]}}' != {'test_data': {'test': ['test_1', 'test_2', 'test_3']}}
> 
> ----------------------------------------------------------------------
> Ran 23 tests in 163.573s
> ```

@afabiani  There are mainly two failing tests from master, this inclusive, However i have fixed them in this commit https://github.com/GeoNode/geonode/pull/7563/commits/705b70f7682d80dd37b22f6681bf77b5f0f3efb2

**Pull request merged**